src/include/txn.i

```diff
@@ -193,6 +193,205 @@ __wt_timestamp_subone(wt_timestamp_t *ts)
+
+	WT_RET(__wt_open_cursor(session,
+	    op->btree->dhandle->name, NULL, open_cursor_cfg, &cursor));
+
+	/*
+	 * Transaction prepare is cleared temporarily as cursor functions are
+	 * not allowed for prepared transactions.
+	 */
+	F_CLR(txn, WT_TXN_PREPARE);
+	if (op->type == WT_TXN_OP_BASIC_ROW ||
+	    op->type == WT_TXN_OP_INMEM_ROW)
+		__wt_cursor_set_raw_key(cursor, &op->u.op_row.key);
+	else
+		((WT_CURSOR_BTREE *)cursor)->iface.recno =
+		    op->u.op_col.recno;
+	F_SET(txn, WT_TXN_PREPARE);
+
+	WT_WITH_BTREE(session,
+	    op->btree, ret = __wt_btcur_search_uncommitted(
+	    (WT_CURSOR_BTREE *)cursor, &upd));
+	WT_ERR(ret);
+
+#ifdef HAVE_DIAGNOSTIC
+	if (upd == NULL)
+		WT_ASSERT(session, txn->pre_processed_count != 0);
+
+	pre_processed_count = txn->pre_processed_count;
+#endif
+
+	op->u.op_upd = upd;
+
+	for (; upd != NULL; upd = upd->next) {
+		 if (upd->txnid != txn->id)
+			continue;
+
+		if (!commit) {
+			upd->txnid = WT_TXN_ABORTED;
+			continue;
+		}
+
+		/*
+		 * Newer updates are inserted at head of update chain, and
+		 * transaction operations are added at the tail of the
+		 * transaction modify chain.
+		 *
+		 * For example, a transaction has modified [k,v] as
+		 * [k, v]  -> [k, u1]   (txn_op : txn_op1)
+		 * [k, u1] -> [k, u2]   (txn_op : txn_op2)
+		 * update chain : u2->u1
+		 * txn_mod      : txn_op1->txn_op2.
+		 *
+		 * Only the key is saved in the transaction operation
+		 * structure, hence we cannot identify whether "txn_op1"
+		 * corresponds to "u2" or "u1" during commit/rollback.
+		 *
+		 * To make things simpler we will handle all the updates
+		 * that match the key saved in a transaction operation in a
+		 * single go. As a result, multiple updates of a key, if any
+		 * will be resolved as part of the first transaction operation
+		 * resolution of that key, and subsequent transaction operation
+		 * resolution of the same key will be effectively
+		 * a no-op.
+		 *
+		 * In the above example, we will resolve "u2" and "u1" as part
+		 * of resolving "txn_op1" and will not do any significant
+		 * thing as part of "txn_op2".
+		 *
+		 * When an update is not identified for resolution of a txn_op,
+		 * we need to distinguish whether the update is already pre
+		 * processed or not. Transaction maintains a reference count to
+		 * keep track of pre-processed updates. Having only one
+		 * reference count, will not solve all cases, but better than
+		 * void.
+		 */
+		if (upd->prepare_state == WT_PREPARE_RESOLVED) {
+#ifdef HAVE_DIAGNOSTIC
+			txn->pre_processed_count--;
+#endif
+			break;
+		}
+
+		/* Resolve the prepared update to be committed update. */
+		__txn_resolve_prepared_update(session, upd);
+#ifdef HAVE_DIAGNOSTIC
+		txn->pre_processed_count++;
+#endif
+	}
+
+#ifdef HAVE_DIAGNOSTIC
+	/*
+	 * Reduce by one, to exclude the current processed one as to keep track
+	 * of only pre-processed ones.
+	 */
+	if (pre_processed_count < txn->pre_processed_count)
+		txn->pre_processed_count--;
```

Comment:
I find this logic really hard to follow (decrementing, then incrementing, now decrementing again).

Could something like this work instead?
```
if (upd != op->u.op_upd) {
    if (upd->prepare_state == WT_PREPARE_RESOLVED) {
        WT_ASSERT(session, txn->multi_update_count > 0);
        --txn->multi_update_count;
    } else
        ++txn->multi_update_count;
}
```